TechnicalDocuments/Proposals/BuildersToFunctions.md

```diff
@@ -0,0 +1,161 @@
+                                 enableClinicalDownloads: Bool,
+                                 addressSwitches: AddressSwitchesProtocol,
+                                 primary: Flow,
+                                 secondary: Flow) -> UIViewController
+    }
+}
+```
+
+At the call site, we would replace this:
+
+```swift
+case let .appointmentList(session):
+    builders
+            .appointmentList(with: session,
+                             enableClinicalDownloads: builders.enableClinicalDownloads,
+                             addressSwitches: builders.addressSwitches)
+            .make(primary: navigation, secondary: modal)
+            |> navigation.present
+```
+
+With this:
+
+```swift
+case let .appointmentList(session):
+    Builder.ClinicalRecords
+            .makeAppointmentList(with: session,
+                                enableClinicalDownloads: enableClinicalDownloads,
+                                addressSwitches: addressSwitches,
+                                primary: navigation,
+                                secondary: modal)
+            |> navigation.present
+```
+
+## Impact on an existing codebase
+
+As Ilya pointed out it would make unit testing a bit more cumbersome. That's the case because you no longer inject builders. I would like to make the case for the following:
+
+1. We don't test `FlowControllers` with Unit tests anyway. This is not a reason to not do it, it's simply difficult to justify time doing so, when there are more important pieces to be tested and those tests would bring, in my opinion, little value. More value would be achieved by the next point (**2.**).
+2. We can test that a particular navigation/flow is done as expected independently of this proposal being approved or not. For that to happen we simply observe a ViewModel state and assert on it, since we derive routes from the state.
+3. If we "really" want to test FlowControllers, it's still possible by mocking `Current`, which is something trivial to achieve (`World.init()`). The difference here is that we can't say we are **Unit** testing FlowControllers, but instead these become **Integration** tests. I personally don't feel this is semantically significant. 
+
+Given all those points, it has no impact, since we don't test that layer and there are currently no plans in the short term to do so. 
+
+There is still the question about moving `UserSession`, or something in that shape, to `Current`. Based on an initial conversation with Anders, and quoting him, it would be a "fairly low" effort. This is for me a point that should be thought carefully and be a reason to reject this proposal. 
+
+As a proposal on its own, it wouldn't have any immediate impact. For new features, and assuming `UserSession` in `Current` is in place, those could start using this approach. Builders that don't rely on `UserSession` could be migrated immediately. 
+
+## Alternatives considered
+
+Leave as is.
```

Comment:
I don't think we should leave things as it is, this issues was brought up multiple times in the past and by ignoring it we don't achieve anything, it's good to see things moving forward with this proposal. But we can maybe explore alternative solutions which will not compromise testability.

Here is a possible alternative to consider:

```swift
extension Builder {
  struct ClinicalRecords {
    var makeMedicalHistory: (_ modal: Flow) -> UIViewController = { /* implementation here */ }
  }
}
```

We don't have to provide any dependencies to this builder because it can access `Current` or just create what is needed in the builder method implementation (at most inject only session into it), but we will be able to override this in unit test and use a view model double instead of a real one. A downside of using closures here is that we can't use named parameters on the call site.

This is similar to pointfree approach, describe [here](https://www.pointfree.co/blog/posts/21-how-to-control-the-world):

```swift
struct API {
  var setToken = { APIClient.shared.token = $0 }
  var fetchCurrentUser = APIClient.shared.fetchCurrentUser
}
```

Reply:
In your suggested approach, where would these Builder structs be created? We need to ensure that the builder methods were immutable in production code.

Reply:
I think they may be created at the same places where they are created now, we just remove constructor parameters from them. Or we may create them where we need them which is in flow controllers, as default values for their arguments:

```swift
class ClinicalRecords {
  init(builder: Builder.ClinicalRecords = Builder.ClinicalRecords())
}
```

Reply:
This is a different design decision/direction and as such, if you would like to propose something else, please feel free to do so in a different place, after there's an outcome for this proposal.

Reply:
I don't see how this would be better than amending this proposal.

Reply:
I am not vouching for that change since it goes in a different direction: it still uses builders as entities that you pass around.

Reply:
> it still uses builders as entities that you pass around.
this is not needed with default initialiser parameters. Approach we already use in `Current`.